### PR TITLE
use polling after submitted

### PR DESCRIPTION
**Pull request by tnsetting**
# Hey, I just made a Pull Request!

## Description
Transition from submitted state to running, terminate state.

## Have you tested this? If so, how?
I have included unit tests.

## Checklist for PR author(s)

- [x] Changes are covered by unit test
- [x] All tests pass
- [ ] Code coverage check passes
- [ ] Error handling is tested
- [ ] Errors are handled at the appropriate layer
- [ ] Errors that cannot be handled where they occur are propagated
- [ ] (optional) Changes are covered by system test
- [ ] Relevant documentation updated
- [x] This PR has NO breaking change to public API
- [ ] This PR has breaking change to public API and it is documented

## Checklist for PR reviewer(s)

- [ ] This PR has been incorporated in release note for the coming version
- [ ] Risky changes introduced by this PR have been all considered

### CODE REVIEW

File: styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/FlyteRunnerHandler.java [hunk 1]
```diff
@@ -108,18 +106,12 @@ private void transitionSubmitting(RunState state, EventRouter eventRouter) {
...
-    final var started = Event.started(state.workflowInstance());
-    LOG.info("Issue 'started' event for: " + state.workflowInstance());
-    sendEventIgnoreClosed(state, eventRouter, "started", started);
-  }
-
-  private void transitionRunning(RunState state, EventRouter eventRouter) {
-    LOG.info("Entered state RUNNING for: " + state.workflowInstance());
+  private void pollingExecution(RunState state, EventRouter eventRouter) {
+    LOG.info("Entered state" + state.state().toString() + " for: " + state.workflowInstance());
     final FlyteExecConf flyteExecConf = state.data().executionDescription().orElseThrow().flyteExecConf().orElseThrow();
```
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/FlyteRunnerHandler.java:111]: I know it is not the problem of this PR, but this method does too much and leaks `FlyteExecConf` that we can probably expose to flyte runner only, plus `flyteRunner.poll` needs `state` anyway.

Comment (honnix) [reply styx-scheduler-service/src/main/java/com/spotify/styx/state/handlers/FlyteRunnerHandler.java:111]: You can ignore this comment as I see the rest of the class does things similar.

File: styx-scheduler-service/src/main/java/com/spotify/styx/flyte/FlyteAdminClientRunner.java [hunk 2]
```diff
@@ -115,6 +115,12 @@ void emitFlyteEvents(ExecutionOuterClass.Execution execution, RunState runState)
     final Execution.WorkflowExecution.Phase phase = execution.getClosure().getPhase();
     final FlytePhase flytePhase = FlytePhase.fromProto(phase);
     switch (flytePhase) {
+      case RUNNING:
+        if (runState.state() == RunState.State.SUBMITTED) {
```
[the remark this answers is not in the document]
Comment (sonjaer) [reply styx-scheduler-service/src/main/java/com/spotify/styx/flyte/FlyteAdminClientRunner.java:119]: Same comment. If we miss to read the execution when it is in the RUNNING phase in flyte, won't we miss the started event. 
I think KubernetesDockerRunner handles this (e.g., [here](https://github.com/spotify/styx/blob/ca09dc516c2ddc7bd163750084aea26955f06680/styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesPodEventTranslator.java#L168)), although not in an optimal way according to this [comment](https://github.com/spotify/styx/blob/master/styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java#L653).

[the remark this answers is not in the document]
Comment (tnsetting) [reply styx-scheduler-service/src/main/java/com/spotify/styx/flyte/FlyteAdminClientRunner.java:119]: Right now the `emitFlyteEvents` is only called under polling which is invoked in `SUBMITTED` and `RUNNING` states. So we only need to distinguish whether the current state is `SUBMITTED` and `RUNNING`.In this case, it is fine to not do anything.

[the remark this answers is not in the document]
Comment (tnsetting) [reply styx-scheduler-service/src/main/java/com/spotify/styx/flyte/FlyteAdminClientRunner.java:119]: @soapraj Aha. I see your point. So when the state transition is from `submitted` to other state, the cases needs to be covered.

[the remark this answers is not in the document]
Comment (sonjaer) [reply styx-scheduler-service/src/main/java/com/spotify/styx/flyte/FlyteAdminClientRunner.java:119]: yes exactly. So maybe add a check for if we're in RunState.State.SUBMITTED in the other later Flyte Phases, send started event, and break. Even though, we'll be polling twice. Not sure if its better than to "spoof" the counter value as is done in the KubernetesDockerRunner

[the remark this answers is not in the document]
Comment (tnsetting) [reply styx-scheduler-service/src/main/java/com/spotify/styx/flyte/FlyteAdminClientRunner.java]: End up with implementing something similar with `KubernetesDockerRunner` which I think would make the code a bit cleaner.

**Comment by tnsetting** [comment]
The `codecov/patch` is failing due to the private constructer of `FlyteEventTranslator` is not covered in the unittest. Talked with @narape and we think it is OK to ignore this error.

File: styx-scheduler-service/src/main/java/com/spotify/styx/flyte/FlyteEventTranslator.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    if (isStarted(flytePhase)) {
+      return handleStarted(flytePhase, runState);
+    }
+    return List.of();
+  }
+
+  static private boolean isExited(FlytePhase flytePhase) {
+    switch (flytePhase) {
```
Comment (narape) [inline styx-scheduler-service/src/main/java/com/spotify/styx/flyte/FlyteEventTranslator.java:58]: I think that it would be better to add a couple of methods to the `FlytePhase` enum, so we can call `flytePhase.isExited()` or `flytePhase.isStarted()` directly instead of these switch.

But we can do that in another PR

File: styx-scheduler-service/src/test/java/com/spotify/styx/flyte/FlyteEventTranslatorTest.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                .newBuilder()
+                .setCode(flyteExitCode)
+                .build())
+            .setPhase(phase).build())
+        .build();
+  }
+
+  @Test
+  public void submittedToTerminate(){
+    assertGeneratesEventsAndTransitions(
```
Comment (narape) [inline styx-scheduler-service/src/test/java/com/spotify/styx/flyte/FlyteEventTranslatorTest.java:86]: Nice, I like it @tnsetting

**Review by narape — APPROVED** [review]

**Comment by narape** [comment]
Here we have a battle between tools. Error prone or whatever wants us to create a private constructor for the utility class `FlyteEventTranslator.java` to prevent instantiation, but then codecov complains that the empty constructor that just throws an exception is not covered by test :shrug:. 

It would be stupid to remove the private visibility and a test to please codecov. So I am merging like it is now

**Comment by honnix** [comment]
We do have a test utility to help test this kind of thing, probably in styx-test module.

**Comment by honnix** [comment]
https://github.com/spotify/styx/blob/master/styx-test/src/main/java/com/spotify/styx/util/ClassEnforcer.java
